Windows: Support sub-millisecond sleep

ChrisDenton opened the pull request:
Use `CreateWaitableTimerExW` with `CREATE_WAITABLE_TIMER_HIGH_RESOLUTION`. Does not work before Windows 10, version 1803 so in that case we fallback to using `Sleep`.

I've created a `WaitableTimer` type so it can one day be adapted to also support waiting to an absolute time (which has been talked about). Note though that it currently returns `Err(())` because we can't do anything with the errors other than fallback to the old `Sleep`. Feel free to tell me to do errors properly. It just didn't seem worth constructing an `io::Error` if we're never going to surface it to the user. And it *should* all be infallible anyway unless the OS is too old to support it.

Closes #43376

ChrisDenton committed:
Windows: Support sub-millisecond sleep

Use `CreateWaitableTimerExW` with `CREATE_WAITABLE_TIMER_HIGH_RESOLUTION`. Does not work before Windows 10, version 1803 so in that case we fallback to using `Sleep`.

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6425489784/job/17448143296) [(plain)](https://github.com/rust-lang/rust/commit/9ac43c5b8477038dcd420a371015d8730faddaee/checks/17448143296/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=ChrisDenton
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_d7a86c24-0032-4b95-a7b7-125b66672985
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=d4f37749298491ecdb1229fcb3a1cbe1cfa14e12
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_d7a86c24-0032-4b95-a7b7-125b66672985
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_d7a86c24-0032-4b95-a7b7-125b66672985
GITHUB_TRIGGERING_ACTOR=ChrisDenton
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/116461/merge
GITHUB_WORKFLOW_SHA=d4f37749298491ecdb1229fcb3a1cbe1cfa14e12
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
+++ <stderr output>
Error: 
   0: tests failed

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
+error: unsupported operation: can't call foreign function `CreateWaitableTimerExW` on OS `windows`
+   |
+   |
+LL | /             c::CreateWaitableTimerExW(
+LL | |                 null(),
+LL | |                 null(),
+LL | |                 c::CREATE_WAITABLE_TIMER_HIGH_RESOLUTION,
+LL | |                 c::TIMER_ALL_ACCESS,
+LL | |             )
+   | |_____________^ can't call foreign function `CreateWaitableTimerExW` on OS `windows`

Caused by:
  process didn't exit successfully: `/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/compiletest-a737b97aade21f89 --quiet` (exit status: 1)
+   |
+   |
+   = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
+   = note: BACKTRACE:
+   = note: inside `std::sys::PLATFORM::time::WaitableTimer::high_resolution` at RUSTLIB/std/src/sys/PLATFORM/time.rs:LL:CC
+   = note: inside `std::sys::PLATFORM::thread::Thread::sleep::high_precision_sleep` at RUSTLIB/std/src/sys/PLATFORM/thread.rs:LL:CC
+   = note: inside `std::sys::PLATFORM::thread::Thread::sleep` at RUSTLIB/std/src/sys/PLATFORM/thread.rs:LL:CC
+   = note: inside `std::thread::sleep` at RUSTLIB/std/src/thread/mod.rs:LL:CC
+  --> $DIR/sleep_long.rs:LL:CC
+   |
+   |
+LL |     thread::sleep(Duration::from_millis(100));
+
+note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace
+
+error: aborting due to previous error
+error: aborting due to previous error
+


error: there were 1 unmatched diagnostics that occurred outside the testfile and had no pattern
    Error: unsupported operation: can't call foreign function `CreateWaitableTimerExW` on OS `windows`
full stderr:
full stderr:
error: unsupported operation: can't call foreign function `CreateWaitableTimerExW` on OS `windows`
   |
   |
LL | /             c::CreateWaitableTimerExW(
LL | |                 null(),
LL | |                 null(),
LL | |                 c::CREATE_WAITABLE_TIMER_HIGH_RESOLUTION,
LL | |                 c::TIMER_ALL_ACCESS,
LL | |             )
   | |_____________^ can't call foreign function `CreateWaitableTimerExW` on OS `windows`
   = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
   = note: BACKTRACE:
   = note: inside `std::sys::windows::time::WaitableTimer::high_resolution` at /checkout/library/std/src/sys/windows/time.rs:236:13: 241:14
   = note: inside `std::sys::windows::time::WaitableTimer::high_resolution` at /checkout/library/std/src/sys/windows/time.rs:236:13: 241:14
   = note: inside `std::sys::windows::thread::Thread::sleep::high_precision_sleep` at /checkout/library/std/src/sys/windows/thread.rs:92:25: 92:57
   = note: inside `std::sys::windows::thread::Thread::sleep` at /checkout/library/std/src/sys/windows/thread.rs:99:37: 99:62
   = note: inside `std::thread::sleep` at /checkout/library/std/src/thread/mod.rs:872:5: 872:28
  --> tests/pass/shims/sleep_long.rs:14:5
   |
   |
LL |     thread::sleep(Duration::from_millis(100));

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace

error: aborting due to previous error
error: aborting due to previous error


full stdout:



FAILED TEST: tests/pass/concurrency/sync.rs (revision `stack`)
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/miri" "--error-format=json" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--target" "i686-pc-windows-msvc" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/tests/pass/concurrency" "tests/pass/concurrency/sync.rs" "--cfg=stack" "-Zmiri-disable-isolation" "-Zmiri-strict-provenance" "--edition" "2021"
error: pass test got exit status: 1, but expected 0

error: actual output differed from expected
Execute `./miri test --bless` to update `tests/pass/concurrency/sync.stack.stderr` to the actual output
Execute `./miri test --bless` to update `tests/pass/concurrency/sync.stack.stderr` to the actual output
--- tests/pass/concurrency/sync.stack.stderr
+++ <stderr output>
+error: unsupported operation: can't call foreign function `CreateWaitableTimerExW` on OS `windows`
+   |
+   |
+LL | /             c::CreateWaitableTimerExW(
+LL | |                 null(),
+LL | |                 null(),
+LL | |                 c::CREATE_WAITABLE_TIMER_HIGH_RESOLUTION,
+LL | |                 c::TIMER_ALL_ACCESS,
+LL | |             )
+   | |_____________^ can't call foreign function `CreateWaitableTimerExW` on OS `windows`
+   = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
+   = note: BACKTRACE:
+   = note: inside `std::sys::PLATFORM::time::WaitableTimer::high_resolution` at RUSTLIB/std/src/sys/PLATFORM/time.rs:LL:CC
+   = note: inside `std::sys::PLATFORM::time::WaitableTimer::high_resolution` at RUSTLIB/std/src/sys/PLATFORM/time.rs:LL:CC
+   = note: inside `std::sys::PLATFORM::thread::Thread::sleep::high_precision_sleep` at RUSTLIB/std/src/sys/PLATFORM/thread.rs:LL:CC
+   = note: inside `std::sys::PLATFORM::thread::Thread::sleep` at RUSTLIB/std/src/sys/PLATFORM/thread.rs:LL:CC
+   = note: inside `std::thread::sleep` at RUSTLIB/std/src/thread/mod.rs:LL:CC
+  --> $DIR/sync.rs:LL:CC
+   |
+   |
+LL |         thread::sleep(Duration::from_millis(200));
+
+note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace
+
+error: aborting due to previous error
+error: aborting due to previous error
+


error: actual output differed from expected
Execute `./miri test --bless` to update `tests/pass/concurrency/sync.stack.stdout` to the actual output
+++ <stdout output>
-before wait
-before wait
-before wait
---
-after wait


error: there were 1 unmatched diagnostics that occurred outside the testfile and had no pattern
    Error: unsupported operation: can't call foreign function `CreateWaitableTimerExW` on OS `windows`
full stderr:
full stderr:
error: unsupported operation: can't call foreign function `CreateWaitableTimerExW` on OS `windows`
   |
   |
LL | /             c::CreateWaitableTimerExW(
LL | |                 null(),
LL | |                 null(),
LL | |                 c::CREATE_WAITABLE_TIMER_HIGH_RESOLUTION,
LL | |                 c::TIMER_ALL_ACCESS,
LL | |             )
   | |_____________^ can't call foreign function `CreateWaitableTimerExW` on OS `windows`
   = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
   = note: BACKTRACE:
   = note: inside `std::sys::windows::time::WaitableTimer::high_resolution` at /checkout/library/std/src/sys/windows/time.rs:236:13: 241:14
   = note: inside `std::sys::windows::time::WaitableTimer::high_resolution` at /checkout/library/std/src/sys/windows/time.rs:236:13: 241:14
   = note: inside `std::sys::windows::thread::Thread::sleep::high_precision_sleep` at /checkout/library/std/src/sys/windows/thread.rs:92:25: 92:57
   = note: inside `std::sys::windows::thread::Thread::sleep` at /checkout/library/std/src/sys/windows/thread.rs:99:37: 99:62
   = note: inside `std::thread::sleep` at /checkout/library/std/src/thread/mod.rs:872:5: 872:28
  --> tests/pass/concurrency/sync.rs:214:9
   |
   |
LL |         thread::sleep(Duration::from_millis(200));

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace

error: aborting due to previous error
error: aborting due to previous error


full stdout:



FAILED TEST: tests/pass/concurrency/sync.rs (revision `tree`)
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/miri" "--error-format=json" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--target" "i686-pc-windows-msvc" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/tests/pass/concurrency" "tests/pass/concurrency/sync.rs" "--cfg=tree" "-Zmiri-disable-isolation" "-Zmiri-strict-provenance" "-Zmiri-tree-borrows" "--edition" "2021"
error: pass test got exit status: 1, but expected 0

error: actual output differed from expected
error: actual output differed from expected
Execute `./miri test --bless` to update `tests/pass/concurrency/sync.tree.stderr` to the actual output
--- tests/pass/concurrency/sync.tree.stderr
+++ <stderr output>
+error: unsupported operation: can't call foreign function `CreateWaitableTimerExW` on OS `windows`
+   |
+   |
+LL | /             c::CreateWaitableTimerExW(
+LL | |                 null(),
+LL | |                 null(),
+LL | |                 c::CREATE_WAITABLE_TIMER_HIGH_RESOLUTION,
+LL | |                 c::TIMER_ALL_ACCESS,
+LL | |             )
+   | |_____________^ can't call foreign function `CreateWaitableTimerExW` on OS `windows`
+   = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
+   = note: BACKTRACE:
+   = note: inside `std::sys::PLATFORM::time::WaitableTimer::high_resolution` at RUSTLIB/std/src/sys/PLATFORM/time.rs:LL:CC
+   = note: inside `std::sys::PLATFORM::time::WaitableTimer::high_resolution` at RUSTLIB/std/src/sys/PLATFORM/time.rs:LL:CC
+   = note: inside `std::sys::PLATFORM::thread::Thread::sleep::high_precision_sleep` at RUSTLIB/std/src/sys/PLATFORM/thread.rs:LL:CC
+   = note: inside `std::sys::PLATFORM::thread::Thread::sleep` at RUSTLIB/std/src/sys/PLATFORM/thread.rs:LL:CC
+   = note: inside `std::thread::sleep` at RUSTLIB/std/src/thread/mod.rs:LL:CC
+  --> $DIR/sync.rs:LL:CC
+   |
+   |
+LL |         thread::sleep(Duration::from_millis(200));
+
+note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace
+
+error: aborting due to previous error
+error: aborting due to previous error
+


error: actual output differed from expected
Execute `./miri test --bless` to update `tests/pass/concurrency/sync.tree.stdout` to the actual output
--- tests/pass/concurrency/sync.tree.stdout
+++ <stdout output>
-before wait
-before wait
-before wait
-before wait
---
-after wait


error: there were 1 unmatched diagnostics that occurred outside the testfile and had no pattern
    Error: unsupported operation: can't call foreign function `CreateWaitableTimerExW` on OS `windows`
full stderr:
full stderr:
error: unsupported operation: can't call foreign function `CreateWaitableTimerExW` on OS `windows`
   |
   |
LL | /             c::CreateWaitableTimerExW(
LL | |                 null(),
LL | |                 null(),
LL | |                 c::CREATE_WAITABLE_TIMER_HIGH_RESOLUTION,
LL | |                 c::TIMER_ALL_ACCESS,
LL | |             )
   | |_____________^ can't call foreign function `CreateWaitableTimerExW` on OS `windows`
   = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
   = note: BACKTRACE:
   = note: inside `std::sys::windows::time::WaitableTimer::high_resolution` at /checkout/library/std/src/sys/windows/time.rs:236:13: 241:14
   = note: inside `std::sys::windows::time::WaitableTimer::high_resolution` at /checkout/library/std/src/sys/windows/time.rs:236:13: 241:14
   = note: inside `std::sys::windows::thread::Thread::sleep::high_precision_sleep` at /checkout/library/std/src/sys/windows/thread.rs:92:25: 92:57
   = note: inside `std::sys::windows::thread::Thread::sleep` at /checkout/library/std/src/sys/windows/thread.rs:99:37: 99:62
   = note: inside `std::thread::sleep` at /checkout/library/std/src/thread/mod.rs:872:5: 872:28
  --> tests/pass/concurrency/sync.rs:214:9
   |
   |
LL |         thread::sleep(Duration::from_millis(200));

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace

error: aborting due to previous error
---
error: actual output differed from expected
Execute `./miri test --bless` to update `tests/pass/shims/time.stderr` to the actual output
--- tests/pass/shims/time.stderr
+++ <stderr output>
+error: unsupported operation: can't call foreign function `CreateWaitableTimerExW` on OS `windows`
+   |
+   |
+LL | /             c::CreateWaitableTimerExW(
+LL | |                 null(),
+LL | |                 null(),
+LL | |                 c::CREATE_WAITABLE_TIMER_HIGH_RESOLUTION,
+LL | |                 c::TIMER_ALL_ACCESS,
+LL | |             )
+   | |_____________^ can't call foreign function `CreateWaitableTimerExW` on OS `windows`
+   = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
+   = note: BACKTRACE:
+   = note: inside `std::sys::PLATFORM::time::WaitableTimer::high_resolution` at RUSTLIB/std/src/sys/PLATFORM/time.rs:LL:CC
+   = note: inside `std::sys::PLATFORM::time::WaitableTimer::high_resolution` at RUSTLIB/std/src/sys/PLATFORM/time.rs:LL:CC
+   = note: inside `std::sys::PLATFORM::thread::Thread::sleep::high_precision_sleep` at RUSTLIB/std/src/sys/PLATFORM/thread.rs:LL:CC
+   = note: inside `std::sys::PLATFORM::thread::Thread::sleep` at RUSTLIB/std/src/sys/PLATFORM/thread.rs:LL:CC
+   = note: inside `std::thread::sleep` at RUSTLIB/std/src/thread/mod.rs:LL:CC
+note: inside `test_sleep`
+   |
+   |
+LL |     std::thread::sleep(Duration::from_millis(100));
+note: inside `main`
+  --> $DIR/time.rs:LL:CC
+   |
+LL |     test_sleep();
---
Build completed unsuccessfully in 0:00:48


error: there were 1 unmatched diagnostics that occurred outside the testfile and had no pattern
    Error: unsupported operation: can't call foreign function `CreateWaitableTimerExW` on OS `windows`
full stderr:
full stderr:
error: unsupported operation: can't call foreign function `CreateWaitableTimerExW` on OS `windows`
   |
   |
LL | /             c::CreateWaitableTimerExW(
LL | |                 null(),
LL | |                 null(),
LL | |                 c::CREATE_WAITABLE_TIMER_HIGH_RESOLUTION,
LL | |                 c::TIMER_ALL_ACCESS,
LL | |             )
   | |_____________^ can't call foreign function `CreateWaitableTimerExW` on OS `windows`
   = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
   = note: BACKTRACE:
   = note: inside `std::sys::windows::time::WaitableTimer::high_resolution` at /checkout/library/std/src/sys/windows/time.rs:236:13: 241:14
   = note: inside `std::sys::windows::time::WaitableTimer::high_resolution` at /checkout/library/std/src/sys/windows/time.rs:236:13: 241:14
   = note: inside `std::sys::windows::thread::Thread::sleep::high_precision_sleep` at /checkout/library/std/src/sys/windows/thread.rs:92:25: 92:57
   = note: inside `std::sys::windows::thread::Thread::sleep` at /checkout/library/std/src/sys/windows/thread.rs:99:37: 99:62
   = note: inside `std::thread::sleep` at /checkout/library/std/src/thread/mod.rs:872:5: 872:28
note: inside `test_sleep`
   |
   |
LL |     std::thread::sleep(Duration::from_millis(100));
note: inside `main`
  --> tests/pass/shims/time.rs:51:5
   |
LL |     test_sleep();
---
error: actual output differed from expected
Execute `./miri test --bless` to update `tests/pass/shims/time-with-isolation.stderr` to the actual output
--- tests/pass/shims/time-with-isolation.stderr
+++ <stderr output>
+error: unsupported operation: can't call foreign function `CreateWaitableTimerExW` on OS `windows`
+   |
+   |
+LL | /             c::CreateWaitableTimerExW(
+LL | |                 null(),
+LL | |                 null(),
+LL | |                 c::CREATE_WAITABLE_TIMER_HIGH_RESOLUTION,
+LL | |                 c::TIMER_ALL_ACCESS,
+LL | |             )
+   | |_____________^ can't call foreign function `CreateWaitableTimerExW` on OS `windows`
+   = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
+   = note: BACKTRACE:
+   = note: inside `std::sys::PLATFORM::time::WaitableTimer::high_resolution` at RUSTLIB/std/src/sys/PLATFORM/time.rs:LL:CC
+   = note: inside `std::sys::PLATFORM::time::WaitableTimer::high_resolution` at RUSTLIB/std/src/sys/PLATFORM/time.rs:LL:CC
+   = note: inside `std::sys::PLATFORM::thread::Thread::sleep::high_precision_sleep` at RUSTLIB/std/src/sys/PLATFORM/thread.rs:LL:CC
+   = note: inside `std::sys::PLATFORM::thread::Thread::sleep` at RUSTLIB/std/src/sys/PLATFORM/thread.rs:LL:CC
+   = note: inside `std::thread::sleep` at RUSTLIB/std/src/thread/mod.rs:LL:CC
+note: inside `test_sleep`
+   |
+   |
+LL |     std::thread::sleep(Duration::from_secs(3600));
+note: inside `main`
+  --> $DIR/time-with-isolation.rs:LL:CC
+   |
+LL |     test_sleep();
---
+


error: there were 1 unmatched diagnostics that occurred outside the testfile and had no pattern
    Error: unsupported operation: can't call foreign function `CreateWaitableTimerExW` on OS `windows`
full stderr:
full stderr:
error: unsupported operation: can't call foreign function `CreateWaitableTimerExW` on OS `windows`
   |
   |
LL | /             c::CreateWaitableTimerExW(
LL | |                 null(),
LL | |                 null(),
LL | |                 c::CREATE_WAITABLE_TIMER_HIGH_RESOLUTION,
LL | |                 c::TIMER_ALL_ACCESS,
LL | |             )
   | |_____________^ can't call foreign function `CreateWaitableTimerExW` on OS `windows`
   = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
   = note: BACKTRACE:
   = note: inside `std::sys::windows::time::WaitableTimer::high_resolution` at /checkout/library/std/src/sys/windows/time.rs:236:13: 241:14
   = note: inside `std::sys::windows::time::WaitableTimer::high_resolution` at /checkout/library/std/src/sys/windows/time.rs:236:13: 241:14
   = note: inside `std::sys::windows::thread::Thread::sleep::high_precision_sleep` at /checkout/library/std/src/sys/windows/thread.rs:92:25: 92:57
   = note: inside `std::sys::windows::thread::Thread::sleep` at /checkout/library/std/src/sys/windows/thread.rs:99:37: 99:62
   = note: inside `std::thread::sleep` at /checkout/library/std/src/thread/mod.rs:872:5: 872:28
note: inside `test_sleep`
   |
   |
LL |     std::thread::sleep(Duration::from_secs(3600));
note: inside `main`
  --> tests/pass/shims/time-with-isolation.rs:43:5
   |
LL |     test_sleep();
```

</details>

ChrisDenton commented:
That error looks like it's just miri again. I'll fix it.

ChrisDenton committed:
Miri: Add `CreateWaitableTimerEx` stub

This function will always fail, allowing std's `Sleep` fallback path to be taken instead.

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6425853877/job/17449148672) [(plain)](https://github.com/rust-lang/rust/commit/6698d22a5c2d95036c0cc9a215047c49aea5b284/checks/17449148672/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=ChrisDenton
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_cafab556-9878-44c1-9304-f13bf936a903
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=a32d34fddd5d9f32af7b400597094a6b9d672797
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_cafab556-9878-44c1-9304-f13bf936a903
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_cafab556-9878-44c1-9304-f13bf936a903
GITHUB_TRIGGERING_ACTOR=ChrisDenton
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/116461/merge
GITHUB_WORKFLOW_SHA=a32d34fddd5d9f32af7b400597094a6b9d672797
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
+++ <stderr output>
Error: 
   0: tests failed

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
+error: unsupported operation: can't call foreign function `CreateWaitableTimerExW` on OS `windows`
+   |
+   |
+LL | /             c::CreateWaitableTimerExW(
+LL | |                 null(),
+LL | |                 null(),
+LL | |                 c::CREATE_WAITABLE_TIMER_HIGH_RESOLUTION,
+LL | |                 c::TIMER_ALL_ACCESS,
+LL | |             )
+   | |_____________^ can't call foreign function `CreateWaitableTimerExW` on OS `windows`
+   = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
+   = note: BACKTRACE:
+   = note: inside `std::sys::PLATFORM::time::WaitableTimer::high_resolution` at RUSTLIB/std/src/sys/PLATFORM/time.rs:LL:CC
+   = note: inside `std::sys::PLATFORM::time::WaitableTimer::high_resolution` at RUSTLIB/std/src/sys/PLATFORM/time.rs:LL:CC
+   = note: inside `std::sys::PLATFORM::thread::Thread::sleep::high_precision_sleep` at RUSTLIB/std/src/sys/PLATFORM/thread.rs:LL:CC
+   = note: inside `std::sys::PLATFORM::thread::Thread::sleep` at RUSTLIB/std/src/sys/PLATFORM/thread.rs:LL:CC
+   = note: inside `std::thread::sleep` at RUSTLIB/std/src/thread/mod.rs:LL:CC
+  --> $DIR/sleep_long.rs:LL:CC
+   |
+   |
+LL |     thread::sleep(Duration::from_millis(100));
+
+note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace
+
+error: aborting due to previous error
+error: aborting due to previous error
+


error: there were 1 unmatched diagnostics that occurred outside the testfile and had no pattern
    Error: unsupported operation: can't call foreign function `CreateWaitableTimerExW` on OS `windows`
full stderr:
full stderr:
error: unsupported operation: can't call foreign function `CreateWaitableTimerExW` on OS `windows`
   |
   |
LL | /             c::CreateWaitableTimerExW(
LL | |                 null(),
LL | |                 null(),
LL | |                 c::CREATE_WAITABLE_TIMER_HIGH_RESOLUTION,
LL | |                 c::TIMER_ALL_ACCESS,
LL | |             )
   | |_____________^ can't call foreign function `CreateWaitableTimerExW` on OS `windows`
   = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
   = note: BACKTRACE:
   = note: inside `std::sys::windows::time::WaitableTimer::high_resolution` at /checkout/library/std/src/sys/windows/time.rs:236:13: 241:14
   = note: inside `std::sys::windows::time::WaitableTimer::high_resolution` at /checkout/library/std/src/sys/windows/time.rs:236:13: 241:14
   = note: inside `std::sys::windows::thread::Thread::sleep::high_precision_sleep` at /checkout/library/std/src/sys/windows/thread.rs:92:25: 92:57
   = note: inside `std::sys::windows::thread::Thread::sleep` at /checkout/library/std/src/sys/windows/thread.rs:99:37: 99:62
   = note: inside `std::thread::sleep` at /checkout/library/std/src/thread/mod.rs:872:5: 872:28
  --> tests/pass/shims/sleep_long.rs:14:5
   |
   |
LL |     thread::sleep(Duration::from_millis(100));

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace

error: aborting due to previous error
error: aborting due to previous error


full stdout:



FAILED TEST: tests/pass/concurrency/sync.rs (revision `stack`)
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/miri" "--error-format=json" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--target" "i686-pc-windows-msvc" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/tests/pass/concurrency" "tests/pass/concurrency/sync.rs" "--cfg=stack" "-Zmiri-disable-isolation" "-Zmiri-strict-provenance" "--edition" "2021"
error: pass test got exit status: 1, but expected 0

error: actual output differed from expected
Execute `./miri test --bless` to update `tests/pass/concurrency/sync.stack.stderr` to the actual output
Execute `./miri test --bless` to update `tests/pass/concurrency/sync.stack.stderr` to the actual output
--- tests/pass/concurrency/sync.stack.stderr
+++ <stderr output>
+error: unsupported operation: can't call foreign function `CreateWaitableTimerExW` on OS `windows`
+   |
+   |
+LL | /             c::CreateWaitableTimerExW(
+LL | |                 null(),
+LL | |                 null(),
+LL | |                 c::CREATE_WAITABLE_TIMER_HIGH_RESOLUTION,
+LL | |                 c::TIMER_ALL_ACCESS,
+LL | |             )
+   | |_____________^ can't call foreign function `CreateWaitableTimerExW` on OS `windows`
+   = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
+   = note: BACKTRACE:
+   = note: inside `std::sys::PLATFORM::time::WaitableTimer::high_resolution` at RUSTLIB/std/src/sys/PLATFORM/time.rs:LL:CC
+   = note: inside `std::sys::PLATFORM::time::WaitableTimer::high_resolution` at RUSTLIB/std/src/sys/PLATFORM/time.rs:LL:CC
+   = note: inside `std::sys::PLATFORM::thread::Thread::sleep::high_precision_sleep` at RUSTLIB/std/src/sys/PLATFORM/thread.rs:LL:CC
+   = note: inside `std::sys::PLATFORM::thread::Thread::sleep` at RUSTLIB/std/src/sys/PLATFORM/thread.rs:LL:CC
+   = note: inside `std::thread::sleep` at RUSTLIB/std/src/thread/mod.rs:LL:CC
+  --> $DIR/sync.rs:LL:CC
+   |
+   |
+LL |         thread::sleep(Duration::from_millis(200));
+
+note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace
+
+error: aborting due to previous error
+error: aborting due to previous error
+


error: actual output differed from expected
Execute `./miri test --bless` to update `tests/pass/concurrency/sync.stack.stdout` to the actual output
+++ <stdout output>
-before wait
-before wait
-before wait
---
-after wait


error: there were 1 unmatched diagnostics that occurred outside the testfile and had no pattern
    Error: unsupported operation: can't call foreign function `CreateWaitableTimerExW` on OS `windows`
full stderr:
full stderr:
error: unsupported operation: can't call foreign function `CreateWaitableTimerExW` on OS `windows`
   |
   |
LL | /             c::CreateWaitableTimerExW(
LL | |                 null(),
LL | |                 null(),

Caused by:
  process didn't exit successfully: `/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/compiletest-a737b97aade21f89 --quiet` (exit status: 1)
LL | |                 c::CREATE_WAITABLE_TIMER_HIGH_RESOLUTION,
LL | |                 c::CREATE_WAITABLE_TIMER_HIGH_RESOLUTION,
LL | |                 c::TIMER_ALL_ACCESS,
LL | |             )
   | |_____________^ can't call foreign function `CreateWaitableTimerExW` on OS `windows`
   = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
   = note: BACKTRACE:
   = note: inside `std::sys::windows::time::WaitableTimer::high_resolution` at /checkout/library/std/src/sys/windows/time.rs:236:13: 241:14
   = note: inside `std::sys::windows::time::WaitableTimer::high_resolution` at /checkout/library/std/src/sys/windows/time.rs:236:13: 241:14
   = note: inside `std::sys::windows::thread::Thread::sleep::high_precision_sleep` at /checkout/library/std/src/sys/windows/thread.rs:92:25: 92:57
   = note: inside `std::sys::windows::thread::Thread::sleep` at /checkout/library/std/src/sys/windows/thread.rs:99:37: 99:62
   = note: inside `std::thread::sleep` at /checkout/library/std/src/thread/mod.rs:872:5: 872:28
  --> tests/pass/concurrency/sync.rs:214:9
   |
   |
LL |         thread::sleep(Duration::from_millis(200));

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace

error: aborting due to previous error
error: aborting due to previous error


full stdout:



FAILED TEST: tests/pass/concurrency/sync.rs (revision `tree`)
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/miri" "--error-format=json" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--target" "i686-pc-windows-msvc" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/tests/pass/concurrency" "tests/pass/concurrency/sync.rs" "--cfg=tree" "-Zmiri-tree-borrows" "-Zmiri-disable-isolation" "-Zmiri-strict-provenance" "--edition" "2021"
error: pass test got exit status: 1, but expected 0

error: actual output differed from expected
error: actual output differed from expected
Execute `./miri test --bless` to update `tests/pass/concurrency/sync.tree.stderr` to the actual output
--- tests/pass/concurrency/sync.tree.stderr
+++ <stderr output>
+error: unsupported operation: can't call foreign function `CreateWaitableTimerExW` on OS `windows`
+   |
+   |
+LL | /             c::CreateWaitableTimerExW(
+LL | |                 null(),
+LL | |                 null(),
+LL | |                 c::CREATE_WAITABLE_TIMER_HIGH_RESOLUTION,
+LL | |                 c::TIMER_ALL_ACCESS,
+LL | |             )
+   | |_____________^ can't call foreign function `CreateWaitableTimerExW` on OS `windows`
+   = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
+   = note: BACKTRACE:
+   = note: inside `std::sys::PLATFORM::time::WaitableTimer::high_resolution` at RUSTLIB/std/src/sys/PLATFORM/time.rs:LL:CC
+   = note: inside `std::sys::PLATFORM::time::WaitableTimer::high_resolution` at RUSTLIB/std/src/sys/PLATFORM/time.rs:LL:CC
+   = note: inside `std::sys::PLATFORM::thread::Thread::sleep::high_precision_sleep` at RUSTLIB/std/src/sys/PLATFORM/thread.rs:LL:CC
+   = note: inside `std::sys::PLATFORM::thread::Thread::sleep` at RUSTLIB/std/src/sys/PLATFORM/thread.rs:LL:CC
+   = note: inside `std::thread::sleep` at RUSTLIB/std/src/thread/mod.rs:LL:CC
+  --> $DIR/sync.rs:LL:CC
+   |
+   |
+LL |         thread::sleep(Duration::from_millis(200));
+
+note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace
+
+error: aborting due to previous error
+error: aborting due to previous error
+


error: actual output differed from expected
Execute `./miri test --bless` to update `tests/pass/concurrency/sync.tree.stdout` to the actual output
--- tests/pass/concurrency/sync.tree.stdout
+++ <stdout output>
-before wait
-before wait
-before wait
-before wait
---
-after wait


error: there were 1 unmatched diagnostics that occurred outside the testfile and had no pattern
    Error: unsupported operation: can't call foreign function `CreateWaitableTimerExW` on OS `windows`
full stderr:
full stderr:
error: unsupported operation: can't call foreign function `CreateWaitableTimerExW` on OS `windows`
   |
   |
LL | /             c::CreateWaitableTimerExW(
LL | |                 null(),
LL | |                 null(),
LL | |                 c::CREATE_WAITABLE_TIMER_HIGH_RESOLUTION,
LL | |                 c::TIMER_ALL_ACCESS,
LL | |             )
   | |_____________^ can't call foreign function `CreateWaitableTimerExW` on OS `windows`
   = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
   = note: BACKTRACE:
   = note: inside `std::sys::windows::time::WaitableTimer::high_resolution` at /checkout/library/std/src/sys/windows/time.rs:236:13: 241:14
   = note: inside `std::sys::windows::time::WaitableTimer::high_resolution` at /checkout/library/std/src/sys/windows/time.rs:236:13: 241:14
   = note: inside `std::sys::windows::thread::Thread::sleep::high_precision_sleep` at /checkout/library/std/src/sys/windows/thread.rs:92:25: 92:57
   = note: inside `std::sys::windows::thread::Thread::sleep` at /checkout/library/std/src/sys/windows/thread.rs:99:37: 99:62
   = note: inside `std::thread::sleep` at /checkout/library/std/src/thread/mod.rs:872:5: 872:28
  --> tests/pass/concurrency/sync.rs:214:9
   |
   |
LL |         thread::sleep(Duration::from_millis(200));

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace

error: aborting due to previous error
---
error: actual output differed from expected
Execute `./miri test --bless` to update `tests/pass/shims/time.stderr` to the actual output
--- tests/pass/shims/time.stderr
+++ <stderr output>
+error: unsupported operation: can't call foreign function `CreateWaitableTimerExW` on OS `windows`
+   |
+   |
+LL | /             c::CreateWaitableTimerExW(
+LL | |                 null(),
+LL | |                 null(),
+LL | |                 c::CREATE_WAITABLE_TIMER_HIGH_RESOLUTION,
+LL | |                 c::TIMER_ALL_ACCESS,
+LL | |             )
+   | |_____________^ can't call foreign function `CreateWaitableTimerExW` on OS `windows`
+   = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
+   = note: BACKTRACE:
+   = note: inside `std::sys::PLATFORM::time::WaitableTimer::high_resolution` at RUSTLIB/std/src/sys/PLATFORM/time.rs:LL:CC
+   = note: inside `std::sys::PLATFORM::time::WaitableTimer::high_resolution` at RUSTLIB/std/src/sys/PLATFORM/time.rs:LL:CC
+   = note: inside `std::sys::PLATFORM::thread::Thread::sleep::high_precision_sleep` at RUSTLIB/std/src/sys/PLATFORM/thread.rs:LL:CC
+   = note: inside `std::sys::PLATFORM::thread::Thread::sleep` at RUSTLIB/std/src/sys/PLATFORM/thread.rs:LL:CC
+   = note: inside `std::thread::sleep` at RUSTLIB/std/src/thread/mod.rs:LL:CC
+note: inside `test_sleep`
+   |
+   |
+LL |     std::thread::sleep(Duration::from_millis(100));
+note: inside `main`
+  --> $DIR/time.rs:LL:CC
+   |
+LL |     test_sleep();
---
+


error: there were 1 unmatched diagnostics that occurred outside the testfile and had no pattern
    Error: unsupported operation: can't call foreign function `CreateWaitableTimerExW` on OS `windows`
full stderr:
full stderr:
error: unsupported operation: can't call foreign function `CreateWaitableTimerExW` on OS `windows`
   |
   |
LL | /             c::CreateWaitableTimerExW(
LL | |                 null(),
LL | |                 null(),
LL | |                 c::CREATE_WAITABLE_TIMER_HIGH_RESOLUTION,
LL | |                 c::TIMER_ALL_ACCESS,
LL | |             )
   | |_____________^ can't call foreign function `CreateWaitableTimerExW` on OS `windows`
   = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
   = note: BACKTRACE:
   = note: inside `std::sys::windows::time::WaitableTimer::high_resolution` at /checkout/library/std/src/sys/windows/time.rs:236:13: 241:14
   = note: inside `std::sys::windows::time::WaitableTimer::high_resolution` at /checkout/library/std/src/sys/windows/time.rs:236:13: 241:14
   = note: inside `std::sys::windows::thread::Thread::sleep::high_precision_sleep` at /checkout/library/std/src/sys/windows/thread.rs:92:25: 92:57
   = note: inside `std::sys::windows::thread::Thread::sleep` at /checkout/library/std/src/sys/windows/thread.rs:99:37: 99:62
   = note: inside `std::thread::sleep` at /checkout/library/std/src/thread/mod.rs:872:5: 872:28
note: inside `test_sleep`
   |
   |
LL |     std::thread::sleep(Duration::from_millis(100));
note: inside `main`
  --> tests/pass/shims/time.rs:51:5
   |
LL |     test_sleep();
---
error: actual output differed from expected
Execute `./miri test --bless` to update `tests/pass/shims/time-with-isolation.stderr` to the actual output
--- tests/pass/shims/time-with-isolation.stderr
+++ <stderr output>
+error: unsupported operation: can't call foreign function `CreateWaitableTimerExW` on OS `windows`
+   |
+   |
+LL | /             c::CreateWaitableTimerExW(
+LL | |                 null(),
+LL | |                 null(),
+LL | |                 c::CREATE_WAITABLE_TIMER_HIGH_RESOLUTION,
+LL | |                 c::TIMER_ALL_ACCESS,
+LL | |             )
+   | |_____________^ can't call foreign function `CreateWaitableTimerExW` on OS `windows`
+   = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
+   = note: BACKTRACE:
+   = note: inside `std::sys::PLATFORM::time::WaitableTimer::high_resolution` at RUSTLIB/std/src/sys/PLATFORM/time.rs:LL:CC
+   = note: inside `std::sys::PLATFORM::time::WaitableTimer::high_resolution` at RUSTLIB/std/src/sys/PLATFORM/time.rs:LL:CC
+   = note: inside `std::sys::PLATFORM::thread::Thread::sleep::high_precision_sleep` at RUSTLIB/std/src/sys/PLATFORM/thread.rs:LL:CC
+   = note: inside `std::sys::PLATFORM::thread::Thread::sleep` at RUSTLIB/std/src/sys/PLATFORM/thread.rs:LL:CC
+   = note: inside `std::thread::sleep` at RUSTLIB/std/src/thread/mod.rs:LL:CC
+note: inside `test_sleep`
+   |
+   |
+LL |     std::thread::sleep(Duration::from_secs(3600));
+note: inside `main`
+  --> $DIR/time-with-isolation.rs:LL:CC
+   |
+LL |     test_sleep();
---
+


error: there were 1 unmatched diagnostics that occurred outside the testfile and had no pattern
    Error: unsupported operation: can't call foreign function `CreateWaitableTimerExW` on OS `windows`
full stderr:
full stderr:
error: unsupported operation: can't call foreign function `CreateWaitableTimerExW` on OS `windows`
   |
   |
LL | /             c::CreateWaitableTimerExW(
LL | |                 null(),
LL | |                 null(),
LL | |                 c::CREATE_WAITABLE_TIMER_HIGH_RESOLUTION,
LL | |                 c::TIMER_ALL_ACCESS,
LL | |             )
   | |_____________^ can't call foreign function `CreateWaitableTimerExW` on OS `windows`
   = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
   = note: BACKTRACE:
   = note: inside `std::sys::windows::time::WaitableTimer::high_resolution` at /checkout/library/std/src/sys/windows/time.rs:236:13: 241:14
   = note: inside `std::sys::windows::time::WaitableTimer::high_resolution` at /checkout/library/std/src/sys/windows/time.rs:236:13: 241:14
   = note: inside `std::sys::windows::thread::Thread::sleep::high_precision_sleep` at /checkout/library/std/src/sys/windows/thread.rs:92:25: 92:57
   = note: inside `std::sys::windows::thread::Thread::sleep` at /checkout/library/std/src/sys/windows/thread.rs:99:37: 99:62
   = note: inside `std::thread::sleep` at /checkout/library/std/src/thread/mod.rs:872:5: 872:28
note: inside `test_sleep`
   |
   |
LL |     std::thread::sleep(Duration::from_secs(3600));
note: inside `main`
  --> tests/pass/shims/time-with-isolation.rs:43:5
   |
LL |     test_sleep();
```

</details>

RalfJung commented on src/tools/miri/src/shims/windows/foreign_items.rs:
```diff
@@ -267,6 +267,14 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
             }
+            "CreateWaitableTimerExW" => {
+                let [_attributes, _name, _flags, _access] =
+                    this.check_shim(abi, Abi::System { unwind: false }, link_name, args)?;
```
Generally we do at least read all the arguments of a shim to make sure they are initialized -- or we mark the shim as "only run in std".

ChrisDenton replied:
Ok, I've added the reads.

RalfJung replied:
Thanks! LGTM.

joboet commented on library/std/src/sys/windows/thread.rs:
```diff
@@ -87,7 +88,17 @@ impl Thread {
     pub fn sleep(dur: Duration) {
-        unsafe { c::Sleep(super::dur2timeout(dur)) }
+        fn high_precision_sleep(dur: Duration) -> Result<(), ()> {
+            let timer = WaitableTimer::high_resolution()?;
```
How good is the performance of `CreateWaitableTimerExW`? If it is slow, it might be better to cache the timer in a thread local.

ChrisDenton replied:
It's around 2,000 nanoseconds on my machine (minus the time it takes to call `QueryPerformanceCounter`). That is well within noise for waiting on my machine. Granted, other hardware may be able to perform better but I'd rather avoid adding a thread local unless or until someone can show it's worth it.

thomcc commented:
Looks fine. My recollection is there's sometimes some other steps to do to get the most accurate possible timing (such as adjusting thread priority before/after sleep), but that probably has negative consequences for power usage.

@bors r+

bors commented:
:pushpin: Commit fbf256716e499d9b2ee15c722c63d90d960f2859 has been approved by `thomcc`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=thomcc fbf256716e499d9b2ee15c722c63d90d960f2859 -->
<!-- homu: {"type":"Approved","sha":"fbf256716e499d9b2ee15c722c63d90d960f2859","approver":"thomcc","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit fbf256716e499d9b2ee15c722c63d90d960f2859 with merge cee6db171d03c22f1cbf98a72dffde858a49fbd8...
<!-- homu: {"type":"BuildStarted","head_sha":"fbf256716e499d9b2ee15c722c63d90d960f2859","merge_sha":"cee6db171d03c22f1cbf98a72dffde858a49fbd8"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6626035982/job/18001837030)
Approved by: thomcc
Pushing cee6db171d03c22f1cbf98a72dffde858a49fbd8 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"thomcc","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6626035982/job/18001837030"},"merge_sha":"cee6db171d03c22f1cbf98a72dffde858a49fbd8"} -->

rust-timer commented:
Finished benchmarking commit (cee6db171d03c22f1cbf98a72dffde858a49fbd8): [comparison URL](https://perf.rust-lang.org/compare.html?start=271dcc1d40b57ad129d88fef1aa7f239308fbfb4&end=cee6db171d03c22f1cbf98a72dffde858a49fbd8&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=271dcc1d40b57ad129d88fef1aa7f239308fbfb4&end=cee6db171d03c22f1cbf98a72dffde858a49fbd8&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=271dcc1d40b57ad129d88fef1aa7f239308fbfb4&end=cee6db171d03c22f1cbf98a72dffde858a49fbd8&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 1.2%  | [1.2%, 1.2%]   | 1     |
| Improvements ✅ <br /> (primary)   | -3.6% | [-3.6%, -3.6%] | 1     |
| Improvements ✅ <br /> (secondary) | -2.0% | [-2.0%, -2.0%] | 1     |
| All ❌✅ (primary)                 | -3.6% | [-3.6%, -3.6%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=271dcc1d40b57ad129d88fef1aa7f239308fbfb4&end=cee6db171d03c22f1cbf98a72dffde858a49fbd8&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -7.1% | [-7.1%, -7.1%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=271dcc1d40b57ad129d88fef1aa7f239308fbfb4&end=cee6db171d03c22f1cbf98a72dffde858a49fbd8&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: missing data
**Artifact size**: 304.37 MiB -> 304.32 MiB (-0.02%)

klensy commented:
Should there be documentation update, describing updated behavior?

ChrisDenton commented:
Hm, this should probably be in the release notes. Not sure if anything changed documentation wise? I intentionally kept the behaviour described there (i.e. zero duration calls `Sleep(0)`).

klensy commented:
https://doc.rust-lang.org/nightly/std/thread/fn.sleep.html
This part.
```
Currently, specifying a zero duration on Unix platforms returns immediately without invoking the underlying [nanosleep](https://linux.die.net/man/2/nanosleep) syscall, whereas on Windows platforms the underlying [Sleep](https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-sleep) syscall is always invoked. If the intention is to yield the current time-slice you may want to use [yield_now](https://doc.rust-lang.org/nightly/std/thread/fn.yield_now.html) instead.
```

ChrisDenton commented:
> Currently, specifying a zero duration on... Windows platforms the underlying [Sleep](https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-sleep) syscall is always invoked.

This should still be true even after this PR
